Frameworks/CoreText/CTFramesetter.mm

```diff
@@ -16,96 +16,55 @@
-
-            CTLineRef line = CTTypesetterCreateLine(static_cast<CTTypesetterRef>(frameSetter->_typesetter.get()), lineRange);
-
-            float ascent = 0.0f, descent = 0.0f, leading = 0.0f;
-            const float width = CTLineGetTypographicBounds(line, &ascent, &descent, &leading);
-            const float lineHeight = ascent - descent + leading;
-
-            if (ret) {
-                CGPoint lineOrigin;
-                switch (alignment) {
-                    case kCTRightTextAlignment:
-                        lineOrigin.x = frameSize.size.width - width;
-                        break;
-                    case kCTCenterTextAlignment:
-                        lineOrigin.x = (frameSize.size.width - width) / 2;
-                        break;
-                    default: // kCTLeftTextAlignment
-                        lineOrigin.x = 0.0f;
-                        break;
-                }
-                lineOrigin.y = y - ascent;
-                [ret->_lines addObject:(id)line];
-                ret->_lineOrigins.push_back(lineOrigin);
-            }
-            [static_cast<_CTLine*>(line) release];
-
-            curIdx = pos;
-            if (width > sizeOut->width) {
-                sizeOut->width = width;
-            }
-
-            sizeOut->height += lineHeight;
-
-            y -= lineHeight;
-        }
+    // Call _DWriteWrapper to get _CTLine object list that makes up this frame
+    _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(frameSetter->_typesetter);
+    CFRange range = CFRangeMake(0, typesetter->_charactersLen);
+    NSArray<_CTLine*>* lines = _dwriteGetLines(typesetter, range, frameSize);
+    if (createFrame) {
+        ret->_lines = lines;
     }
 
-    if (ret) {
-        ret->_totalSize = *sizeOut;
+    if (sizeOut) {
+        // TODO::
+        // Is there a better way to do this - Investigate.
+        // For now the below logic is ued to calculate the frame size required to fit the text in -
+        //     width = frame width as it is the constraint passed to DWrite to calculate renderable area width
+        //     height = min(_lineOrigin.y of the last line in the frame + the _lineOrigin.y of the first line (so we have some buffer
+        //              space after the last line), frameSize.size.height)
+
+        // Calculate width
+        sizeOut->width = frameSize.size.width;
+
+        // Calculate heigth
+        _CTLine* firstLine = static_cast<_CTLine*>([lines objectAtIndex:0]);
+        _CTLine* lastLine = static_cast<_CTLine*>([lines objectAtIndex:[lines count] - 1]);
+        // TODO::
+        // Modify this code once we move away from Cairo Freetype for rasterizing text.
+        CGFloat originYFirstLine = frameSize.size.height - firstLine->_lineOrigin.y - frameSize.origin.y;
+        CGFloat originYLastLine = frameSize.size.height - lastLine->_lineOrigin.y - frameSize.origin.y;
+        sizeOut->height = min(originYLastLine + originYFirstLine, frameSize.size.height);
```

Comment:
It cannot be  firstLine->_lineOrigin.y - lastLine->_lineOrigin.y as firstLine can be equal to lastLine if we only have one line.

That said I think my comment above and the code here does not match. Will fix that

---
In reply to: [77259329](https://github.com/Microsoft/WinObjC/pull/950#discussion_r77259329) [](ancestors = 77259329)